pilot/pkg/model/push_context.go

```diff
@@ -1663,54 +1660,6 @@ func (ps *PushContext) initMeshNetworks() {
 
 }
 
-func (ps *PushContext) initClusterLocalHosts(e *Environment) {
```

Comment:
I thought about that, but it would require a more substantial refactor than (in my opinion) we want to build into this PR. `PushContext.IsClusterLocal` is used by both the [cluster_builder](https://github.com/istio/istio/blob/master/pilot/pkg/networking/core/v1alpha3/cluster_builder.go#L304) and [endpoint_builder](https://github.com/istio/istio/blob/master/pilot/pkg/xds/endpoint_builder.go#L242) and, as far as I can tell, neither of those have a trivial way to leverage the environment for this information.

I'm happy to do a bigger refactor here if you think it's the appropriate way to move forward, but that was my rationale for duplicating the clusterLocalHosts info in both the push context and the environment.

Reply:
Understood. Could we at least move the population of `clusterLocalHosts` to the environment and have it copied over when we create the push context? I think that should work.

Reply:
I believe that's what I have done [here](https://github.com/istio/istio/pull/29763/files#diff-a8e332e06b003dfbf58c29f37de893fbb0e45f642200ea3ea2faa9847fc7e1e0R915) - when `pushContext.InitContext` is called, it calls out to the environment to generate the cluster local hosts, which updates the env list as well as the one in the newly initialized pushcontext

Reply:
Ah you're right. The part that's missing is the bit about `initMeshHandlers`. We don't want to re-initialize `clusterLocalHosts` each time we create a `PushContext`. Instead, we only want to re-initialize it when the MeshConfig changes. To do that, we need to add logic to `initMeshHandlers` to call `env.initClusterLocalHosts`. Then, when we create a `PushContext`, we just copy over the value from the `Environment`.